controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java

```diff
@@ -696,49 +696,65 @@ public void listStreamsInScope(Controller.StreamsInScopeRequest request, StreamO
-
-                                            boolean isAuthorized = grpcAuthHelper.isAuthorized(streamAuthResource,
-                                                    AuthHandler.Permissions.READ, ctx);
-                                            log.debug("Authorization for [{}] for READ permission was [{}]",
-                                                    streamAuthResource, isAuthorized);
-                                            return isAuthorized;
-                                        })
-                                        .map(m -> StreamInfo.newBuilder().setScope(scopeName).setStream(m).build())
-                                        .collect(Collectors.toList());
                                 return Controller.StreamsInScopeResponse
-                                        .newBuilder().addAllStreams(streams)
-                                        .setContinuationToken(Controller.ContinuationToken.newBuilder()
-                                                                .setToken(response.getValue()).build())
+                                        .newBuilder().addAllStreams(response.getKey())
+                                        .setContinuationToken(Controller.ContinuationToken.newBuilder().setToken(response.getValue()).build())
                                         .setStatus(Controller.StreamsInScopeResponse.Status.SUCCESS).build();
                             }
-                        }), responseObserver, requestTag);
+                        });
+
+        authenticateExecuteAndProcessResults(
+                () -> {
+                        String result = this.grpcAuthHelper.checkAuthorization(
+                                AuthResourceRepresentation.ofScope(scopeName),
+                                AuthHandler.Permissions.READ,
+                                ctx);
+                        log.debug("Result of authorization for [{}] and READ permission is: [{}]",
+                            AuthResourceRepresentation.ofScope(scopeName), result);
+                        return result;
+                }, streamsFn, responseObserver, requestTag);
+    }
+
+    private CompletableFuture<Pair<List<StreamInfo>, String>> listStreamsInternal(String scope, String continuationToken, int limit, AuthContext ctx) {
+        List<StreamInfo> streams = new ArrayList<>();
+        return controllerService.listStreams(scope, continuationToken, limit)
+                                .thenCompose(response -> {
+                                    log.debug("All streams in scope with continuation token: {}", response);
+                                    // filter unauthorized scopes. 
+                                    // fetch recursively if scopes are filtered out.
+                                    AtomicInteger filteredCount = new AtomicInteger();
+
+                                    response.getKey().stream().filter(x -> {
```

Comment:
Perhaps putting it this way may make it easier to read and we also won't need the Atomic variable
```
List<String> streamNames = response.getKey().stream().filter(x -> grpcAuthHelper.isAuthorized(AuthResourceRepresentation.ofStreamInScope(scope, x), AuthHandler.Permissions.READ, ctx)).collect(Collectors.toList())
if (streamNames.size() < response.getKey()) {
//recurse...
} else {
//return ...
}
```